File: algorithms/graph/hamiltonian-path/go/hamiltonian_path.go
package hamiltonianpath

// HamiltonianPath returns 1 if a Hamiltonian path exists, 0 otherwise.
func HamiltonianPath(arr []int) int {
	n, m := arr[0], arr[1]
	if n <= 1 { return 1 }
	adj := make([][]bool, n)
	for i := range adj { adj[i] = make([]bool, n) }
	for i := 0; i < m; i++ {
		u, v := arr[2+2*i], arr[3+2*i]
		adj[u][v] = true; adj[v][u] = true
	}
	full := (1 << uint(n)) - 1
	dp := make([][]bool, 1<<uint(n))
	for i := range dp { dp[i] = make([]bool, n) }
	for i := 0; i < n; i++ { dp[1<<uint(i)][i] = true }
	for mask := 1; mask <= full; mask++ {
		for i := 0; i < n; i++ {
			if !dp[mask][i] { continue }
			for j := 0; j < n; j++ {
				if mask&(1<<uint(j)) == 0 && adj[i][j] {
					dp[mask|(1<<uint(j))][j] = true
				}
			}
		}
	}
	for i := 0; i < n; i++ { if dp[full][i] { return 1 } }
	return 0
}

File: algorithms/graph/hamiltonian-path/python/hamiltonian_path.py
def hamiltonian_path(arr: list[int]) -> int:
    n = arr[0]
    m = arr[1]
    if n <= 1:
        return 1
    adj = [[False] * n for _ in range(n)]
    for i in range(m):
        u, v = arr[2 + 2 * i], arr[3 + 2 * i]
        adj[u][v] = True
        adj[v][u] = True

    full = (1 << n) - 1
    dp = [[False] * n for _ in range(1 << n)]
    for i in range(n):
        dp[1 << i][i] = True

    for mask in range(1, 1 << n):
        for i in range(n):
            if not dp[mask][i]:
                continue
            for j in range(n):
                if mask & (1 << j) == 0 and adj[i][j]:
                    dp[mask | (1 << j)][j] = True

    for i in range(n):
        if dp[full][i]:
            return 1
    return 0
